pkg/dyninst/irgen/irgen.go

```diff
@@ -1794,24 +1796,95 @@ func populateEventExpressions(
 		default:
 			panic(fmt.Sprintf("unexpected event kind: %v", event.Kind))
 		}
-		variableSize := variable.Type.GetByteSize()
-		expr := &ir.RootExpression{
-			Name:   variable.Name,
-			Offset: uint32(0),
-			Kind:   variableKind,
-			Expression: ir.Expression{
-				Type: variable.Type,
-				Operations: []ir.ExpressionOp{
-					&ir.LocationOp{
-						Variable: variable,
-						Offset:   0,
-						ByteSize: uint32(variableSize),
-					},
-				},
-			},
-		}
+		expr := newRootExpressionForVariable(variable, variableKind)
 		expressions = append(expressions, expr)
 	}
+
+	// Process template segments if template exists
+	// This is to
+	var variableExpressionSet map[string]int
+	if probe.Template != nil {
+		// Build a set of existing variable expressions for quick lookup
+		variableExpressionSet = make(map[string]int, len(expressions))
+		for i, expr := range expressions {
+			variableExpressionSet[expr.Name] = i
+		}
+
+		for i, segment := range probe.Template.Segments {
+			jsonSeg, ok := segment.(ir.JSONSegment)
+			if !ok {
+				continue // Skip string segments
+			}
+
+			// Parse the DSL expression
+			expr, err := exprlang.Parse(jsonSeg.JSON)
+			if err != nil {
+				return ir.Issue{
+					Kind:    ir.IssueKindUnsupportedFeature,
+					Message: fmt.Sprintf("failed to parse template segment: %v", err),
+				}
+			}
+
+			var refExpr *exprlang.RefExpr
+			switch e := expr.(type) {
+			case *exprlang.RefExpr:
+				refExpr = e
+			default:
+				return ir.Issue{
+					Kind:    ir.IssueKindUnsupportedFeature,
+					Message: fmt.Sprintf("unsupported template operation: %T, only 'ref' operation is supported", expr),
+				}
+			}
+
+			// Find the variable in the subprogram
+			var targetVariableRole ir.VariableRole
+			switch event.Kind {
+			case ir.EventKindEntry:
+				targetVariableRole = ir.VariableRoleParameter
+			case ir.EventKindReturn:
+				targetVariableRole = ir.VariableRoleReturn
+			case ir.EventKindLine:
+				targetVariableRole = ir.VariableRoleLocal
+			default:
+				panic(fmt.Sprintf("unexpected event kind: %v", event.Kind))
+			}
+
+			var targetVar *ir.Variable
+			for _, v := range probe.Subprogram.Variables {
+				if v.Role != targetVariableRole {
+					continue
+				}
+				if v.Name == refExpr.Ref {
+					targetVar = v
+					break
+				}
+			}
+			if targetVar == nil {
+				// Variable not found - could be unavailable at this event kind
+				continue
+			}
+
+			jsonSeg.EventKind = event.Kind
+			// Check if we already have an expression for this variable
+			if existingIdx, ok := variableExpressionSet[targetVar.Name]; ok {
+				// Reuse existing expression index
+				jsonSeg.EventExpressionIndex = existingIdx
+				probe.Template.Segments[i] = jsonSeg
+				continue
+			}
```

Comment:
this is an optimization -- maybe we skip it for the first pass here? I feel like it confuses the story